File: ax_telemetry.py
"""
Aarvex Portal Telemetry — client error / health reports + admin Command Center.

Storage (DynamoDB, same table):
  pk = TELEMETRY#{YYYY-MM-DD}#{event_id}
  ttl  = 30 days

Portal:  POST /telemetry/report   (optional auth, rate-limited)
Admin:   GET  /admin/api/telemetry
Admin:   POST /admin/api/telemetry/ack
Admin:   GET  /admin/api/setup-map
"""

from __future__ import annotations

import json
import logging
import time
import uuid
from datetime import datetime, timezone
from decimal import Decimal

logger = logging.getLogger()

_HELPERS = {}


def _bind(helpers: dict) -> None:
    _HELPERS.update(helpers or {})


def _now() -> str:
    return datetime.now(timezone.utc).isoformat()


def _ttl(days: int = 30) -> int:
    return int(time.time()) + days * 86400


def _json_response(status: int, body: dict):
    fn = _HELPERS.get("json_response")
    if fn:
        return fn(status, body)
    return {
        "statusCode": status,
        "headers": {
            "Content-Type": "application/json",
            "Access-Control-Allow-Origin": "*",
            "Access-Control-Allow-Headers": "Content-Type,Authorization,X-Admin-Session",
            "Access-Control-Allow-Methods": "GET,POST,OPTIONS",
        },
        "body": json.dumps(body, default=str),
    }


def _table():
    return _HELPERS.get("table")


def _event_body(event: dict) -> dict:
    fn = _HELPERS.get("event_body")
    if fn:
        return fn(event) or {}
    raw = event.get("body") or "{}"
    if event.get("isBase64Encoded"):
        import base64
        raw = base64.b64decode(raw).decode("utf-8")
    try:
        return json.loads(raw) if isinstance(raw, str) else (raw or {})
    except Exception:
        return {}


def _admin_ok(event: dict) -> bool:
    fn = _HELPERS.get("admin_authorized")
    return bool(fn and fn(event))


def _rate_limited(key: str, action: str, limit: int = 30, window: int = 60) -> bool:
    fn = _HELPERS.get("rate_limited")
    if not fn:
        return False
    try:
        return bool(fn(key, action, limit, window))
    except TypeError:
        try:
            return bool(fn(key, action))
        except Exception:
            return False
    except Exception:
        return False


# ── Static setup map (source of truth for Admin Command Center) ──────────
SETUP_MAP = {
    "version": 1,
    "updated": "2026-08-06",
    "tabs": [
        {
            "id": "dashboard",
            "label": "Home",
            "features": ["banner", "search_people", "stories", "feed", "geo", "messages_fab", "profile_alert"],
            "files": ["portal.html", "ax-feed.js", "ax-stories.js", "ax-social.js"],
            "design_rules": ["solid_notif_bell", "msg_fab_no_dark_circle", "story_static_ring", "search_edge_flush"],
        },
        {
            "id": "trade",
            "label": "Trade",
            "features": ["banner", "search_products", "category_grid", "product_sections", "shops_mode", "cart", "cod_online"],
            "files": ["portal-catalogue-render.js", "product-modal.js", "portal-order-flow.js", "catalogue-ui.css"],
            "design_rules": ["category_rounded_boxes", "search_opens_category_grid", "shop_category_sections_3col"],
        },
        {
            "id": "favourites",
            "label": "Favourites",
            "features": ["fav_products", "fav_shops", "category_grid"],
            "files": ["ax-feed.js", "product-modal.js"],
            "design_rules": ["fav_shops_same_as_trade_shop_profile"],
        },
        {
            "id": "delivery",
            "label": "Delivery",
            "features": ["claims", "live_map", "otp_complete", "earnings"],
            "files": ["ax-delivery-ux.js", "portal-delivery.js"],
            "design_rules": [],
        },
        {
            "id": "account",
            "label": "Account",
            "features": ["orders_history", "invoice", "payouts", "delete_account"],
            "files": ["portal-order-flow.js", "portal-auth.js"],
            "design_rules": [],
        },
    ],
    "global_features": [
        {"id": "preloader", "files": ["ax-preloader.js", "assets/car-loader.json"], "notes": "Line-art + blur host"},
        {"id": "messenger", "files": ["ax-messenger.js"], "notes": "In-app chat + calls"},
        {"id": "telemetry", "files": ["ax-telemetry.js", "ax_telemetry.py"], "notes": "Client health + admin CC"},
        {"id": "cod_invoice", "files": ["cod_invoice.py"], "notes": "Invoice after OTP; online on payment"},
    ],
    "safe_autofix": [
        "ignore_benign_browser_noise",
        "remount_lottie_preloader",
        "clear_stale_docked_search",
        "trim_local_error_buffer",
        "retry_once_chunk_load",
    ],
    "never_autofix_without_permission": [
        "payment_flow",
        "order_placement",
        "invoice_generation",
        "auth_session",
        "account_delete",
        "delivery_otp",
        "admin_privileges",
        "schema_migrations",
    ],
}


def handle_telemetry_report(event: dict) -> dict:
    """Portal → cloud: batch of client events."""
    body = _event_body(event)
    events = body.get("events") or []
    if isinstance(body.get("event"), dict):
        events = [body["event"]]
    if not isinstance(events, list) or not events:
        return _json_response(400, {"error": "No events"})

    # Rate limit by coarse IP / session fingerprint
    headers = event.get("headers") or {}
    ip = headers.get("x-forwarded-for") or headers.get("X-Forwarded-For") or "anon"
    ip = str(ip).split(",")[0].strip()[:64]
    if _rate_limited(ip or "anon", "telemetry_report"):
        return _json_response(429, {"error": "Too many telemetry reports"})

    table = _table()
    if table is None:
        return _json_response(503, {"error": "Storage unavailable"})

    saved = 0
    day = datetime.now(timezone.utc).strftime("%Y-%m-%d")
    for raw in events[:25]:
        if not isinstance(raw, dict):
            continue
        eid = str(raw.get("id") or uuid.uuid4())[:48]
        severity = str(raw.get("severity") or "info").lower()
        if severity not in ("ignore", "info", "warn", "error", "critical"):
            severity = "error"
        item = {
            "pk": f"TELEMETRY#{day}#{eid}",
            "sk": "EVENT",
            "entity": "telemetry",
            "event_id": eid,
            "day": day,
            "ts": raw.get("ts") or _now(),
            "severity": severity,
            "auto_action": str(raw.get("auto_action") or "none")[:40],
            "auto_applied": str(raw.get("auto_applied") or "none")[:40],
            "tab": str(raw.get("tab") or "unknown")[:40],
            "feature": str(raw.get("feature") or "unknown")[:60],
            "type": str(raw.get("type") or "client")[:40],
            "message": str(raw.get("message") or "")[:500],
            "cause": str(raw.get("cause") or "")[:400],
            "stack": str(raw.get("stack") or "")[:1200],
            "href": str(raw.get("href") or "")[:300],
            "meta": json.dumps(raw.get("meta") or {})[:1500],
            "status": "open",
            "needs_permission": bool(raw.get("needs_permission")),
            "critical": bool(raw.get("critical") or severity == "critical"),
            "ttl": _ttl(30),
            "created_at": _now(),
        }
        try:
            table.put_item(Item=item)
            saved += 1
        except Exception as e:
            logger.warning("[TELEMETRY] put failed: %s", e)

    return _json_response(200, {"ok": True, "saved": saved})


def handle_admin_telemetry_list(event: dict) -> dict:
    if not _admin_ok(event):
        return _json_response(403, {"error": "Unauthorized"})
    table = _table()
    if table is None:
        return _json_response(503, {"error": "Storage unavailable"})

    qs = event.get("queryStringParameters") or {}
    day = (qs.get("day") or datetime.now(timezone.utc).strftime("%Y-%m-%d")).strip()
    limit = min(int(qs.get("limit") or 100), 200)

    items = []
    try:
        # Prefer day prefix query via scan (table may be pk-only)
        from boto3.dynamodb.conditions import Attr
        resp = table.scan(
            FilterExpression=Attr("entity").eq("telemetry") & Attr("day").eq(day),
            Limit=400,
        )
        items = resp.get("Items") or []
        while resp.get("LastEvaluatedKey") and len(items) < 400:
            resp = table.scan(
                FilterExpression=Attr("entity").eq("telemetry") & Attr("day").eq(day),
                ExclusiveStartKey=resp["LastEvaluatedKey"],
                Limit=400,
            )
            items.extend(resp.get("Items") or [])
    except Exception as e:
        logger.warning("[TELEMETRY] list scan failed: %s", e)
        # Fallback: pk prefix scan helper if bound
        scan = _HELPERS.get("scan_by_pk_prefix")
        if scan:
            items = scan(f"TELEMETRY#{day}#") or []

    items.sort(key=lambda x: x.get("ts", ""), reverse=True)
    items = items[:limit]

    by_tab = {}
    by_sev = {}
    open_critical = 0
    for it in items:
        t = it.get("tab") or "unknown"
        by_tab[t] = by_tab.get(t, 0) + 1
        s = it.get("severity") or "info"
        by_sev[s] = by_sev.get(s, 0) + 1
        if it.get("critical") and it.get("status") == "open":
            open_critical += 1

    # Decimal → native for JSON
    def _clean(o):
        if isinstance(o, list):
            return [_clean(x) for x in o]
        if isinstance(o, dict):
            return {k: _clean(v) for k, v in o.items()}
        if isinstance(o, Decimal):
            return int(o) if o % 1 == 0 else float(o)
        return o

    return _json_response(200, {
        "day": day,
        "total": len(items),
        "open_critical": open_critical,
        "by_tab": by_tab,
        "by_severity": by_sev,
        "events": _clean(items),
        "setup_version": SETUP_MAP.get("version"),
    })


def handle_admin_telemetry_ack(event: dict) -> dict:
    if not _admin_ok(event):
        return _json_response(403, {"error": "Unauthorized"})
    body = _event_body(event)
    pk = body.get("pk") or ""
    status = body.get("status") or "acked"
    if not pk.startswith("TELEMETRY#"):
        return _json_response(400, {"error": "Invalid pk"})
    table = _table()
    if table is None:
        return _json_response(503, {"error": "Storage unavailable"})
    try:
        try:
            table.update_item(
                Key={"pk": pk, "sk": "EVENT"},
                UpdateExpression="SET #s = :s, acked_at = :t",
                ExpressionAttributeNames={"#s": "status"},
                ExpressionAttributeValues={":s": status, ":t": _now()},
            )
        except Exception:
            table.update_item(
                Key={"pk": pk},
                UpdateExpression="SET #s = :s, acked_at = :t",
                ExpressionAttributeNames={"#s": "status"},
                ExpressionAttributeValues={":s": status, ":t": _now()},
            )
        return _json_response(200, {"ok": True})
    except Exception as e:
        return _json_response(500, {"error": str(e)})


def handle_admin_setup_map(event: dict) -> dict:
    if not _admin_ok(event):
        return _json_response(403, {"error": "Unauthorized"})
    return _json_response(200, {"setup": SETUP_MAP})


def handle_admin_telemetry_purge(event: dict) -> dict:
    """Delete telemetry rows older than N days (or all for a day)."""
    if not _admin_ok(event):
        return _json_response(403, {"error": "Unauthorized"})
    table = _table()
    if table is None:
        return _json_response(503, {"error": "Storage unavailable"})
    body = _event_body(event)
    days = int(body.get("days") or 30)
    dry_run = bool(body.get("dry_run", True))
    scan = _HELPERS.get("scan_by_pk_prefix")
    items = scan("TELEMETRY#") if scan else []
    cutoff = time.time() - days * 86400
    deleted = 0
    samples = []
    for it in items:
        pk = it.get("pk", "")
        if dry_run:
            if len(samples) < 30:
                samples.append(pk)
            deleted += 1
            continue
        try:
            table.delete_item(Key={"pk": pk})
            deleted += 1
        except Exception:
            pass
    return _json_response(200, {"ok": True, "dry_run": dry_run, "deleted": deleted, "samples": samples, "cutoff_days": days})

File: ax-telemetry.js
/* ═══════════════════════════════════════════════════════════════════════
   ax-telemetry.js — Portal Health + Command Center client
   - Captures errors with tab/feature context
   - Ignores benign noise
   - Safe auto-fixes ONLY (no payment/order/auth changes without permission)
   - Cloud sync → POST /telemetry/report
   - Local buffer for Admin (same-origin) + health checks
   ═══════════════════════════════════════════════════════════════════════ */
(function () {
  var KEY = 'ax_client_error_log_v1';
  var QUEUE_KEY = 'ax_telemetry_queue_v1';
  var MAX = 100;
  var flushTimer = null;

  var SETUP = {
    version: 1,
    tabs: {
      dashboard: { label: 'Home', features: ['banner', 'search_people', 'stories', 'feed', 'geo', 'messages_fab'] },
      trade: { label: 'Trade', features: ['search_products', 'category_grid', 'product_sections', 'shops_mode', 'cart'] },
      favourites: { label: 'Favourites', features: ['fav_products', 'fav_shops'] },
      delivery: { label: 'Delivery', features: ['claims', 'live_map', 'otp'] },
      account: { label: 'Account', features: ['orders', 'invoice'] },
      profile: { label: 'Profile', features: ['kyc', 'address'] }
    },
    neverAutofix: ['payment', 'order', 'invoice', 'auth', 'otp', 'account_delete', 'admin']
  };

  function load(key) {
    try { return JSON.parse(localStorage.getItem(key) || '[]'); } catch (e) { return []; }
  }
  function save(key, arr) {
    try { localStorage.setItem(key, JSON.stringify((arr || []).slice(0, MAX))); } catch (e) {
      try { localStorage.removeItem(key); } catch (e2) {}
    }
  }

  function activeTab() {
    // Portal keeps dynamic per-panel tabs; other pages set window.AX_TAB_HINT.
    var p = document.querySelector('.portal-panel.active');
    if (p && p.id) return String(p.id).replace(/^panel-/, '') || (window.AX_TAB_HINT || 'unknown');
    return window.AX_TAB_HINT || 'unknown';
  }

  function guessFeature(msg, stack) {
    var t = (msg + ' ' + stack).toLowerCase();
    if (/lottie|preloader|car-loader/.test(t)) return 'preloader';
    if (/story|ax-story/.test(t)) return 'stories';
    if (/feed|ax-feed/.test(t)) return 'feed';
    if (/catalogue|productgrid|category/.test(t)) return 'catalogue';
    if (/search|dock|ax-search/.test(t)) return 'search';
    if (/messenger|chat|webrtc/.test(t)) return 'messenger';
    if (/payment|cashfree|razorpay|invoice|cod/.test(t)) return 'checkout';
    if (/delivery|otp|geo/.test(t)) return 'delivery';
    return 'general';
  }

  function isBenign(msg, stack) {
    var t = String(msg || '') + ' ' + String(stack || '');
    return /ResizeObserver loop|Script error\.|chrome-extension:|moz-extension:|safari-extension:|Blocked a frame|Loading CSS chunk|Non-Error promise rejection|AbortError|The operation was aborted|ResizeObserver loop limit exceeded|idb-backend|__gCrWeb/i.test(t);
  }

  function classify(msg, stack, type) {
    var m = String(msg || '');
    var s = String(stack || '');
    var feature = guessFeature(m, s);
    var tab = activeTab();

    if (isBenign(m, s)) {
      return {
        severity: 'ignore', critical: false, needs_permission: false,
        auto_action: 'ignore', cause: 'Benign browser/extension noise — ignored.',
        tab: tab, feature: feature
      };
    }

    // Safe auto-fixable small issues
    if (/lottie|car-loader|ax-lottie|dotlottie/i.test(m + s)) {
      return {
        severity: 'warn', critical: false, needs_permission: false,
        auto_action: 'remount_lottie', cause: 'Preloader animation failed — safe remount.',
        tab: tab, feature: 'preloader'
      };
    }
    if (/QuotaExceeded|localStorage/i.test(m)) {
      return {
        severity: 'warn', critical: false, needs_permission: false,
        auto_action: 'trim_buffer', cause: 'Storage full — trim local telemetry buffer.',
        tab: tab, feature: feature
      };
    }
    if (/ChunkLoadError|Loading chunk [\d]+ failed|dynamically imported module/i.test(m + s)) {
      return {
        severity: 'warn', critical: false, needs_permission: false,
        auto_action: 'reload_once', cause: 'Stale JS chunk — one safe reload allowed.',
        tab: tab, feature: feature
      };
    }

    // Small syntax / missing-symbol issues: soft reload only — NEVER rewrite code
    if (/SyntaxError|Unexpected token/i.test(m)) {
      return {
        severity: 'error', critical: true, needs_permission: true,
        auto_action: 'reload_once', cause: 'Syntax/parse error — one soft reload; code changes need permission.',
        tab: tab, feature: feature
      };
    }
    if (/is not defined|ReferenceError/i.test(m)) {
      return {
        severity: 'error', critical: true, needs_permission: true,
        auto_action: 'reload_once', cause: 'Missing symbol (cache/deploy?) — soft reload; code changes need permission.',
        tab: tab, feature: feature
      };
    }
    if (/Failed to fetch|NetworkError|net::ERR|Load failed|503|502/i.test(m + s)) {
      return {
        severity: 'error', critical: true, needs_permission: true,
        auto_action: 'none', cause: 'API/network failure — investigate Lambda/CORS; no silent code rewrite.',
        tab: tab, feature: feature
      };
    }
    if (/payment|cashfree|razorpay|invoice|otp|order place|account\/delete/i.test(m + s + ' ' + feature)) {
      return {
        severity: 'critical', critical: true, needs_permission: true,
        auto_action: 'none', cause: 'Checkout/auth/delivery critical path — changes require explicit permission.',
        tab: tab, feature: feature
      };
    }
    if (/Cannot read prop|undefined is not|null is not|TypeError/i.test(m)) {
      return {
        severity: 'error', critical: false, needs_permission: true,
        auto_action: 'none', cause: 'Null access — likely race/UI before data; report only.',
        tab: tab, feature: feature
      };
    }

    return {
      severity: 'error', critical: false, needs_permission: true,
      auto_action: 'none', cause: 'Unclassified client error — logged for Admin Command Center.',
      tab: tab, feature: feature
    };
  }

  function pushLocal(entry) {
    var arr = load(KEY);
    arr.unshift(entry);
    save(KEY, arr);
    try { document.dispatchEvent(new CustomEvent('ax:client-error', { detail: entry })); } catch (e) {}
  }

  function enqueueCloud(entry) {
    if (entry.severity === 'ignore') return;
    var q = load(QUEUE_KEY);
    q.push(entry);
    save(QUEUE_KEY, q);
    scheduleFlush();
  }

  function scheduleFlush() {
    if (flushTimer) return;
    flushTimer = setTimeout(function () {
      flushTimer = null;
      flushCloud();
    }, 1800);
  }

  function flushCloud() {
    var q = load(QUEUE_KEY);
    if (!q.length) return;
    var batch = q.slice(0, 20);
    save(QUEUE_KEY, q.slice(20));
    var base = (typeof LAMBDA_URL !== 'undefined' && LAMBDA_URL) ? LAMBDA_URL : '';
    if (!base) return;
    var headers = { 'Content-Type': 'application/json' };
    try {
      var tok = localStorage.getItem('ax_google_token');
      if (tok) headers.Authorization = 'Bearer ' + tok;
    } catch (e) {}
    fetch(base + '/telemetry/report', {
      method: 'POST',
      headers: headers,
      body: JSON.stringify({ events: batch }),
      keepalive: true
    }).catch(function () {
      // put back on failure
      var again = batch.concat(load(QUEUE_KEY));
      save(QUEUE_KEY, again);
    });
  }

  /** Only these run automatically. Payment/order/auth code is NEVER rewritten. */
  var SAFE_AUTO = {
    ignore: 1,
    remount_lottie: 1,
    trim_buffer: 1,
    clear_stale_dock: 1,
    reload_once: 1
  };

  function runSafeAutofix(action) {
    if (!action || action === 'none' || !SAFE_AUTO[action]) return 'skipped';
    if (action === 'ignore') return 'ignored';
    try {
      if (action === 'remount_lottie') {
        document.querySelectorAll('.ax-lottie-host').forEach(function (h) {
          try {
            h.removeAttribute('data-ax-mounted');
            while (h.firstChild) h.removeChild(h.firstChild);
            var wrap = h.closest('.ax-lottie-wrap');
            if (wrap) wrap.classList.remove('is-playing');
          } catch (e) {}
        });
        if (typeof axMountLoaders === 'function') axMountLoaders(document);
        return 'remount_lottie';
      }
      if (action === 'trim_buffer') {
        save(KEY, load(KEY).slice(0, 30));
        save(QUEUE_KEY, load(QUEUE_KEY).slice(0, 20));
        return 'trim_buffer';
      }
      if (action === 'clear_stale_dock') {
        document.querySelectorAll('.portal-panel:not(.active) .catalogue-search-sticky.is-docked').forEach(function (el) {
          el.classList.remove('is-docked');
        });
        return 'clear_stale_dock';
      }
      if (action === 'reload_once') {
        var flag = 'ax_chunk_reload_once';
        if (!sessionStorage.getItem(flag)) {
          sessionStorage.setItem(flag, '1');
          location.reload();
          return 'reload_once';
        }
        return 'reload_suppressed';
      }
    } catch (e) {
      return 'autofix_failed';
    }
    return 'unknown';
  }

  function report(err, meta) {
    var msg = (err && err.message) ? err.message : String(err || 'Error');
    var stack = (err && err.stack) ? err.stack : ((meta && meta.stack) || '');
    var type = (meta && meta.type) || 'manual';
    var info = classify(msg, stack, type);

    // Periodic UI hygiene (safe)
    runSafeAutofix('clear_stale_dock');

    // Soft recovery only from SAFE_AUTO whitelist — big issues stay report-only
    var applied = 'none';
    if (info.auto_action && SAFE_AUTO[info.auto_action]) {
      applied = runSafeAutofix(info.auto_action);
    }

    var entry = {
      id: 'e_' + Date.now().toString(36) + '_' + Math.random().toString(36).slice(2, 7),
      ts: new Date().toISOString(),
      type: type,
      message: String(msg).slice(0, 500),
      stack: String(stack).slice(0, 1200),
      cause: info.cause,
      critical: !!info.critical,
      severity: info.severity,
      needs_permission: !!info.needs_permission,
      auto_action: info.auto_action,
      auto_applied: applied,
      tab: info.tab,
      feature: info.feature,
      href: (location && location.href) || '',
      ua: (navigator && navigator.userAgent) || '',
      meta: meta || null
    };

    if (info.severity !== 'ignore') pushLocal(entry);
    enqueueCloud(entry);
    return entry;
  }

  /* ── Health checks (design + logic signals) ───────────────────────── */
  function runHealthChecks() {
    var findings = [];
    function add(sev, feature, message, cause, needsPerm) {
      findings.push({
        severity: sev, feature: feature, message: message, cause: cause,
        needs_permission: !!needsPerm, critical: sev === 'critical'
      });
    }

    // Story ring should not spin
    var spinning = false;
    try {
      document.querySelectorAll('.ax-story-ring.unseen').forEach(function (el) {
        var an = getComputedStyle(el).animationName || '';
        if (an && an !== 'none') spinning = true;
      });
    } catch (e) {}
    if (spinning) add('warn', 'stories', 'Story ring still animating', 'Design rule: static IG/WA ring only', true);

    // Trade search should open category grid on focus wiring
    var tradeInp = document.getElementById('catalogueSearchInput');
    if (tradeInp && !/openCategoryQuickGrid/.test(String(tradeInp.getAttribute('onfocus') || ''))) {
      add('error', 'search', 'Trade search missing category-grid onfocus', 'Restore onfocus=openCategoryQuickGrid(trade)', true);
    }

    // Docked bar under inactive panel
    if (document.querySelector('.portal-panel:not(.active) .catalogue-search-sticky.is-docked')) {
      add('warn', 'search', 'Stale docked search on hidden panel', 'Safe autofix clear_stale_dock', false);
      runSafeAutofix('clear_stale_dock');
    }

    // Preloader asset presence (best-effort)
    // Shop modal empty products while open — soft signal
    var shopTitle = document.querySelector('.shop-products-title');
    if (shopTitle && /Products \(0\)/.test(shopTitle.textContent || '')) {
      add('info', 'shop_profile', 'Shop profile shows 0 products', 'Check API shop products / category_name', true);
    }

    findings.forEach(function (f) {
      if (f.severity === 'info') return;
      report(f.message, { type: 'health_check', stack: '', feature: f.feature });
    });
    return findings;
  }

  // Public API (compat with ax-error-log.js)
  window.axLogError = function (err, meta) { return report(err, meta); };
  window.axGetClientErrors = function () { return load(KEY); };
  window.axClearClientErrors = function () { save(KEY, []); save(QUEUE_KEY, []); };
  window.axGetSetupMap = function () { return SETUP; };
  window.axRunHealthChecks = runHealthChecks;
  window.axFlushTelemetry = flushCloud;

  window.addEventListener('error', function (e) {
    report(e.error || e.message, {
      type: 'window.onerror',
      source: e.filename,
      line: e.lineno,
      col: e.colno,
      stack: e.error && e.error.stack
    });
  });
  window.addEventListener('unhandledrejection', function (e) {
    report(e.reason || 'Unhandled rejection', { type: 'unhandledrejection' });
  });

  document.addEventListener('DOMContentLoaded', function () {
    setTimeout(runHealthChecks, 3500);
    setInterval(function () { runSafeAutofix('clear_stale_dock'); }, 15000);
    setInterval(flushCloud, 20000);
  });

  window.addEventListener('online', flushCloud);
  document.addEventListener('visibilitychange', function () {
    if (document.visibilityState === 'hidden') flushCloud();
  });
})();
